examples/htool_i2c.c

```diff
@@ -0,0 +1,226 @@
+    }
+    printf("\n");
+  }
+
+  return 0;
+}
+
+static int i2c_read(struct libhoth_device *dev,
+                    const struct htool_invocation *inv) {
+  uint32_t bus;
+  uint32_t freq;
+  uint32_t addr;
+  uint32_t offset;
+  uint32_t length;
+
+  if (htool_get_param_u32(inv, "bus", &bus) ||
+      htool_get_param_u32(inv, "frequency", &freq) ||
+      htool_get_param_u32(inv, "address", &addr) ||
+      htool_get_param_u32(inv, "offset", &offset) ||
+      htool_get_param_u32(inv, "length", &length)) {
+    return -1;
+  }
+
+  struct ec_request_i2c_transfer request;
+  request.bus_number = (uint8_t)(bus & 0xFF);
+  request.dev_address = (uint8_t)(addr & 0x7F);
+  request.size_read = (uint16_t)(length & 0xFFFF);
+  request.speed_khz = (uint16_t)(freq & 0xFFFF);
+  request.flags = 0;
+
+  if (offset == UINT32_MAX) {
+    request.size_write = 0;
+  } else {
+    request.size_write = 1;
+    request.arg_bytes[1] = (offset & 0xFF);
+  }
+
+  uint8_t response[sizeof(struct ec_response_i2c_transfer)];
+  size_t rLen = 0;
+  int ret = htool_exec_hostcmd(
+      dev, EC_CMD_BOARD_SPECIFIC_BASE + EC_PRV_CMD_HOTH_I2C_TRANSFER, 0,
+      &request, sizeof(request), &response, sizeof(response), &rLen);
+  if (ret != 0) {
+    fprintf(stderr, "HOTH_I2C_TRANSFER [read] error code: %d\n", ret);
+    return -1;
+  }
+  if (rLen != sizeof(response)) {
+    fprintf(stderr,
+            "HOTH_I2C_TRANSFER [read] expected exactly %ld response "
+            "bytes, got %ld\n",
+            sizeof(response), rLen);
+    return -1;
+  }
+
+  struct ec_response_i2c_transfer *pI2t =
+      (struct ec_response_i2c_transfer *)(response);
+
+  printf("Read %u bytes from 0x%02x\n", pI2t->read_bytes, request.dev_address);
+  for (uint16_t i = 0; i < pI2t->read_bytes; i++) {
+    printf("0x%02X ", pI2t->resp_bytes[i]);
+  }
+  printf("\n");
+
+  return 0;
+}
+
+static int i2c_write(struct libhoth_device *dev,
+                     const struct htool_invocation *inv) {
+  uint32_t bus;
+  uint32_t freq;
+  uint32_t addr;
+  bool no_stop;
+  char *byte_stream = NULL;
+
+  if (htool_get_param_u32(inv, "bus", &bus) ||
+      htool_get_param_u32(inv, "frequency", &freq) ||
+      htool_get_param_u32(inv, "address", &addr) ||
+      htool_get_param_bool(inv, "no_stop", &no_stop) ||
+      htool_get_param_string(inv, "byte_stream", (const char **)&byte_stream)) {
+    return -1;
+  }
+
+  if (byte_stream == NULL) {
+    return -1;
+  }
+
+  struct ec_request_i2c_transfer request;
+  request.bus_number = (uint8_t)(bus & 0xFF);
+  request.dev_address = (uint8_t)(addr & 0x7F);
+  request.speed_khz = (uint16_t)(freq & 0xFFFF);
+  request.flags = (no_stop ? I2C_BITS_NO_STOP : 0) | I2C_BITS_WRITE;
+  request.size_read = 0;
+
+  uint16_t idx = 0;
+  char *tk = strtok(byte_stream, " ");
+  while (tk && (idx < I2C_TRANSFER_DATA_MAX_SIZE_BYTES)) {
+    char *endptr;
+    unsigned long int parsed = strtoul(tk, &endptr, 0);
+    if ((errno != 0 && parsed == 0) || tk == endptr) {
```

Comment:
According to the documentation for [`strtoul`](https://en.cppreference.com/w/c/string/byte/strtoul), the function only sets `errno`  when the converted integer is outside the range of `unsigned long`, and in that case, `parsed` would be `ULONG_MAX` instead of `0`. The `tk == endptr` already checks if the conversion is valid.

I would include two more checks:

1. Check that `parsed` is actually a byte. `"0x123"` should fail.
2. Check that we parsed the entire input. `"0x"` and `"1a"` should fail.

```suggestion
    if (tk == endptr || parsed > UINT8_MAX || *endptr != '\0') {
```

Reply:
ACK. I misunderstood the way the documentation was worded; errno is set and the result is ULONG_MAX or 0 depending on if a conversion can be completed or not. The documentation for `strtol` is clearer.

I have applied your suggested change (and fixed an array index error I saw after proofing)

Reply:
Thanks!